server/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java

```diff
@@ -417,12 +417,19 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
         changed |= rewriteClauses(queryRewriteContext, mustNotClauses, newBuilder::mustNot);
         changed |= rewriteClauses(queryRewriteContext, filterClauses, newBuilder::filter);
         changed |= rewriteClauses(queryRewriteContext, shouldClauses, newBuilder::should);
+        // early termination when must clause is empty and optional clauses is returning MatchNoneQueryBuilder
+        if(mustClauses.size() == 0 && filterClauses.size() == 0
```

Comment:
Given test failures, I think the problem is that this rewrites empty boolean queries to a match_none, while this is currently treated as a match_all. We probably to do something like below, can you add a unit test for this case?

```suggestion
        if (mustClauses.size() == 0 && filterClauses.size() == 0 && shouldClauses.size() > 0
```

Reply:
Thanks for the catch. I just updated the code. Let me know.